replication-api-impl/src/main/java/org/codice/ditto/replication/api/impl/MetacardConfigLoader.java

```diff
@@ -54,97 +54,80 @@
 
   private final MetacardHelper helper;
 
-  private final MetacardType configMetacardType;
-
   private final SiteManager siteManager;
 
+  private final ReplicatorConfigManager newConfigManager;
+
+  private final Security security;
+
   public MetacardConfigLoader(
       CatalogFramework framework,
       FilterBuilder filterBuilder,
       MetacardHelper helper,
-      MetacardType configMetacardType,
-      SiteManager siteManager) {
+      SiteManager siteManager,
+      ReplicatorConfigManager newConfigManager) {
+    this(framework, filterBuilder, helper, siteManager, newConfigManager, Security.getInstance());
+  }
+
+  @VisibleForTesting
+  MetacardConfigLoader(
+      CatalogFramework framework,
+      FilterBuilder filterBuilder,
+      MetacardHelper helper,
+      SiteManager siteManager,
+      ReplicatorConfigManager newConfigManager,
+      Security security) {
     this.framework = framework;
     this.filterBuilder = filterBuilder;
     this.helper = helper;
-    this.configMetacardType = configMetacardType;
     this.siteManager = siteManager;
+    this.newConfigManager = newConfigManager;
+    this.security = security;
   }
 
-  public ReplicatorConfig createConfig() {
-    return new ReplicatorConfigImpl();
+  public void init() {
+    Failsafe.with(
+            new RetryPolicy()
+                .withBackoff(3, 60, TimeUnit.SECONDS)
+                .withMaxDuration(10, TimeUnit.MINUTES))
+        .onFailedAttempt(
+            throwable -> LOGGER.debug("failed to upgrade legacy configs, will retry", throwable))
+        .onFailure(
+            throwable ->
+                LOGGER.debug("failed to upgrade legacy configs, restart to try again.", throwable))
+        .run(this::convertConfigs);
```

Comment:
❓ How would Failsafe detect errors and retry? From the `convertConfigs()` method, it seems to be returning successfully in all situations.

Reply:
Based on the examples I've seen this seems to retry when any exception is thrown. When running unit tests at one point I was getting a lot of NPEs and the unit tests that ran through failsafe would hang for 10 minutes so I beleive they were hanging because they kept seeing NPEs as well.

Reply:
That being said. On one side, Failsafe would, as you stated, retry on all failures. There are some type of exceptions (i.e. `Throwable`) that should never be ignored. Those are the `VirtualMachineError`. No one should ignore those and continue running as they should be left to bubble out. Failsafe doesn't do that properly and as such, it should not be used with its default configuration. Secondly, as stated below, your `convertConfigs()` method catch exceptions and return successfully with `null` which would let Failsafe believe you were successful.